File: backend_loyalty/backend_loyalty/models/userChallenge.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');
const User = require('./user'); 
const Challenge = require('./challenge'); 

const UserChallenge = sequelize.define('UserChallenge', {
  id: {
    type: DataTypes.BIGINT.UNSIGNED,
    primaryKey: true,
    autoIncrement: true
  },
  status: {
    type: DataTypes.STRING,
    allowNull: true,
  },
  complete_date: {
    type: DataTypes.DATE,
    allowNull: true,
  },
  user_id: {
    type: DataTypes.BIGINT.UNSIGNED,
    references: {
      model: 'users',
      key: 'id'
    }
  },
  challenge_id: {
    type: DataTypes.BIGINT.UNSIGNED,
    references: {
      model: 'challenges', 
      key: 'id'
    }
  },
  created_at: {
    type: DataTypes.DATE,
    allowNull: true,
  },
  updated_at: {
    type: DataTypes.DATE,
    allowNull: true,
  },
}, {
  timestamps: false, 
  paranoid: true,
  tableName: 'user_challenges',
});

UserChallenge.belongsTo(User, { foreignKey: 'user_id', as: 'users', onDelete: 'CASCADE' });
UserChallenge.belongsTo(Challenge, { foreignKey: 'challenge_id', as: 'challenges', onDelete: 'CASCADE' });

module.exports = UserChallenge;

File: loyalty cashier/loyality_cashier/lib/model/user_challenge.dart


import 'package:loyality_cashier/model/challenge.dart';

class UserChallenge {
  final String ?status;
  final DateTime? completeDate;
  final int? userId;
  final int? challengeId;
  final int ?points;
  final Challenge? challenges;

  UserChallenge({
  this.status,
 this.completeDate,
    this.userId,
    this.challengeId,
    this.points,
    this.challenges,
  });

 factory UserChallenge.fromJson(Map<String, dynamic> json) {
  return UserChallenge(
    status: json['status'] ?? '',
    completeDate: json['complete_date'] != null ? DateTime.parse(json['complete_date']) : null,
    userId: json['user_id'] ?? 0,
    challengeId: json['challenge_id'] ?? 0,
    points: json['points'] ?? 0, 
    challenges: json['challenges'] != null ? Challenge.fromJson(json['challenges']) : null,
  );
}


  

  Map<String, dynamic> toJson() {
    return {
      'status': status,
      'complete_date': completeDate?.toIso8601String(),
      'user_id': userId,
      'challenge_id': challengeId,
      'points': points,
      'challenges': challenges?.toJson(),
    };
  }
}
